src/main/php/lang/ClassLoader.class.php

```diff
@@ -42,29 +42,50 @@
     $delegates = [],
     $modules   = [];
 
-  static function __static() {
-    $modules= [];
-    
-    // Scan include-path, setting up classloaders for each element
+  /**
+   * Scan include-path, setting up classloaders for each unique element
+   *
+   * @return void
+   */
+  public static function bootstrap() {
     foreach (\xp::$classpath as $element) {
-      if (DIRECTORY_SEPARATOR === $element{strlen($element) - 1}) {
-        $cl= FileSystemClassLoader::instanceFor($element, false);
-      } else {
-        $cl= ArchiveClassLoader::instanceFor($element, false);
+      if (isset(self::$delegates[$element])) {
+        continue;
+      } else if (is_dir($element)) {
+        self::$delegates[$element]= FileSystemClassLoader::instanceFor($element, false);
+      } else if (is_file($element)) {
+        self::$delegates[$element]= ArchiveClassLoader::instanceFor($element, false);
       }
-      if (isset(self::$delegates[$cl->instanceId()])) continue;
-
-      self::$delegates[$cl->instanceId()]= $cl;
-      if ($cl->providesResource('module.xp')) $modules[]= $cl;
     }
 
-    // Initialize modules
+    // Run deferred initialization
     \xp::$loader= new self();
-    foreach ($modules as $cl) {
-      self::$modules[$cl->instanceId()]= Module::register(self::declareModule($cl));
+    foreach (self::$delegates as $id => $delegate) {
+      self::initialize($id, $delegate);
     }
   }
-  
+
+  /**
+   * Initialize a classloader delegate: Loads autoload.php if not already
+   * required (because loaded via Composer).
+   *
+   * @param  string $id
+   * @param  lang.IClassLoader $cl
+   */
+  private static function initialize($id, $cl) {
+    if ($cl->providesResource('/autoload.php')) {
+      require_once($cl->getResourceAsStream('/autoload.php')->getURI());
+    }
+
+    // BC: If autoload.php did not define a module, check module.xp
+    if (!isset(self::$modules[$id])) {
+      self::$modules[$id]= Module::$INCOMPLETE;
+      if ($cl->providesResource('module.xp')) {
+        self::$modules[$id]= Module::register(self::declareModule($cl));
+      }
+    }
```

Comment:
This BC handling will be removed in XP9 in phase 2; can then start adding `initialize()` to class loader and removing this helper for performance gains.